File: test/selmer/filter_parser_test.py
from __future__ import annotations

from selmer.filter_parser import compile_filter_body, escape_html_star, split_value
from selmer.util import get_accessor, parse_accessor


def test_get_accessor() -> None:
    assert get_accessor([1], 0) == 1
    assert get_accessor({"foo": 1}, "foo") == 1
    assert get_accessor({":foo": 1}, "foo") == 1
    assert parse_accessor("foo.bar.0.baz") == ["foo", "bar", 0, "baz"]
    assert parse_accessor("some..namespace/keyword") == ["some.namespace/keyword"]


def test_split_value_respects_quotes() -> None:
    assert split_value('foo|join:", "|upper') == ["foo", 'join:", "', "upper"]


def test_compile_filter_body() -> None:
    compiled = compile_filter_body("foo.bar|upper")
    assert compiled({"foo": {"bar": "baz"}}) == "BAZ"


def test_escape_html_star() -> None:
    assert escape_html_star("&\"'<>") == "&amp;&quot;&#39;&lt;&gt;"

File: test/selmer/filter-parser.test.ts
import { describe, expect, it } from "vitest";
import { compileFilterBody, escapeHtmlStar, splitValue } from "../../src/filter-parser.js";
import { getAccessor, parseAccessor } from "../../src/util.js";

describe("filter parser helpers", () => {
  it("gets nested values with string and numeric keys", () => {
    expect(getAccessor([1], 0)).toBe(1);
    expect(getAccessor({ foo: 1 }, "foo")).toBe(1);
    expect(getAccessor({ ":foo": 1 }, "foo")).toBe(1);
    expect(parseAccessor("foo.bar.0.baz")).toEqual(["foo", "bar", 0, "baz"]);
    expect(parseAccessor("some..namespace/keyword")).toEqual(["some.namespace/keyword"]);
  });

  it("splits filter bodies while respecting quoted delimiters", () => {
    expect(splitValue('foo|join:", "|upper')).toEqual(["foo", 'join:", "', "upper"]);
  });

  it("compiles filter bodies", () => {
    const compiled = compileFilterBody("foo.bar|upper");
    expect(compiled({ foo: { bar: "baz" } })).toBe("BAZ");
  });

  it("escapes the Django HTML characters", () => {
    expect(escapeHtmlStar("&\"'<>")) .toBe("&amp;&quot;&#39;&lt;&gt;");
  });
});
